### ISSUE TITLE
Cookie Example

### ISSUE THREAD

**Issue by elemprod** [gotev/android-upload-service#94]
Thanks for all the hard work on this project! 

I've been trying to add a cookie to the header which my server requires.   

.addHeader("Cookie", cookieString)

cookieString has the value:

.AspNet.ApplicationCookie=pxhoCVIWJ5eEcEr2fdB5B3DUIzp8cl48VRGE4UlB-kW46sma.....

Do you see anything obviously wrong with or happen to have an example of cookie handling?

Thanks,
Ben

**Comment by gotev** [comment]
Hi!
Android Upload Service uses Android's native `HttpUrlConnection` to perform requests. To use cookies, all you have to do is place the following two lines:
```java
CookieManager cookieManager = new CookieManager();
CookieHandler.setDefault(cookieManager);
```
somewhere before any of your network requests. I would put that into the `Application` subclass. This way cookie management should be done automatically for you and you don't have to set them manually in the requests. Check the [official documentation](http://developer.android.com/reference/java/net/HttpURLConnection.html) (`Sessions with cookies` paragraph). Try and let me know.

**Comment by rohannexialabs** [comment]
Hi, I also want to add session cookie in the request. I am using OkHttp client along with persistent storage to store the cookie. Below is the code - 

 persistentCookieStore = new PersistentCookieStore(ctx);
        // Using OkHTTPClient for managing the cookies received from the server
        OkHttpClient client = new OkHttpClient();
        client.setCookieHandler(new CookieManager(
                persistentCookieStore,
                CookiePolicy.ACCEPT_ALL));

        // Forcing Volley to use OkHttpClient instead of OkHttp
        mRequestQueue = Volley.newRequestQueue(ctx, new OkHttpStack(client));

How should I send session cookies in this scenario?

**Comment by gotev** [comment]
@rohannexialabs your question is a little bit OT, but if I was in your shoes I would have done the following:

My server supports HTTP/2 and SPDY and I'm targeting Android API < 19?
* `Yes`, then find a way to get the cookies associated to the domain of the upload server from the OkHttp Cookie Manager, and manually set the `Cookie` header with the string value got from there in the upload requests
* `No`, get rid of the OkHttp stack underneath Volley and use standard HurlStack, as OkHttp is the default implementation used in Android 4.4+ (API >= 19) to provide `HttpUrlConnection`. Then use CookieManager as described in one of the preceding posts. Cookie management will be done automatically.

**Comment by rohannexialabs** [comment]
Hi, thanks for quick reply. I have tried solution 1st and tried to get the session id and add it to the headers but somehow server isn't picking it. Can you please take a look -

```
String uploadId =
                    new MultipartUploadRequest(context, TransportConstants.URL_FILE)
                            .addFileToUpload(uri.getPath(), "myfile")
                            .addParameter("tdir", current_path)
                            .setMethod("POST")
                            .addHeader("webpy_session_id", sessionId)
                            .setNotificationConfig(new UploadNotificationConfig())
                            .setMaxRetries(2)
                            .startUpload();
```

**Comment by gotev** [comment]
Of course it's not working, you should be setting cookie header like this:
```java
.addHeader("Cookie", "webpy_session_id=" + sessionId)
```
If you have multiple cookies, then separate them with `;`, for example:
```
cookie1=somevalue;cookie2=someothervalue
```
also, `setMethod("POST")` is useless here as it's the default HTTP method used in Multipart Upload. JavaDocs [here](http://gotev.github.io/android-upload-service/javadoc/net/gotev/uploadservice/MultipartUploadRequest.html#setMethod-java.lang.String-).

**Comment by rohannexialabs** [comment]
Thanks for the prompt reply, I have used the code suggested by you but still the authentication issue persists.

**Comment by gotev** [comment]
@rohannexialabs This doesn't seem to me to be a bug in Android Upload Service, as all you do is set a header with a value. If the server does not recognize the session, then you're passing the wrong value. What I suggest you to do in this case, instead of blindlessly trying solutions, is to make a dump with Wireshark and capture:
* the response to the authentication request you make to your server to start the session. You should see a `Set-Cookie` header in the response
* the http multipart request you make to your server with android upload service. Check if the `Cookie` header contains the cookie expected by your server (the one which is contained in the `Set-Cookie` header)

Also, log the parameters you pass into the requests.

**Comment by rohannexialabs** [comment]
Ok will try, thanks for response again.

**Comment by rohannexialabs** [comment]
Your suggestion of using Wireshark was marvelous. Turns out that I was sending wrong session Id, its working fine now. Thanks for quick support and awesome library.

**Comment by gotev** [comment]
Fine! Wireshark is always your best ally when something is not working as expected :)
If you could post a snippet which shows how to properly get the cookie from OkHttp and set that into the upload request, I will include that in the Wiki as a recipe for future reference.

**Comment by rohannexialabs** [comment]
Will be glad to return the favor, might even help me in future :D

```
String uploadId =
                    new MultipartUploadRequest(context, "api url")
                            .addFileToUpload("path to file", "parameter name")
                            .addParameter("addtional parameter name", "value")
                            .addHeader("Cookie", "cookie name=" + sessionId)
                            .setNotificationConfig(new UploadNotificationConfig())
                            .setMaxRetries(2)
                            .startUpload();
```

**Comment by gotev** [comment]
@rohannexialabs Thank you. Could you also post the code snippet with which you get `sessionId` out of OkHttp's Cookie Manager?

**Comment by rohannexialabs** [comment]
Ok sure, but this will vary for different user since the implementation might be a bit different.  I am using `PersistentCookieStore` by `Fran Montiel` (https://gist.github.com/franmontiel/ed12a2295566b7076161). To get the cookie information I created a function in my network singleton class where I have used PersistentCookieStore -  

```
public String getSessionCookie() {
        if (persistentCookieStore != null) {
            for (final HttpCookie cookie : persistentCookieStore.getCookies()) {
                if (cookie.getName().equalsIgnoreCase(Constants.SESSION_COOKIE_NAME)) {
                    return cookie.getValue();
                }
            }
        }
        return null;
    }
```

I called the function from the Activity where I am using your repository and got the session token. I have already posted how I am then using it further.

**Comment by gotev** [comment]
Added a section in the Wiki, with both examples: https://github.com/gotev/android-upload-service/wiki/Advanced-topics#cookies

@rohannexialabs I had to modify your code a little bit to make it generic and usable outside your context

**Comment by elemprod** [comment]
Hi Alex,

Thank you very much for the feedback!   This helped solve my issue.

-Ben

**Comment by gotev** [comment]
@elemprod @rohannexialabs with the just released 2.1 release, the library natively supports OkHttp stack, so workarounds are not needed anymore to manage cookies. Yoi can simply pass your `OkHttpClient` instance when you initialize the library. More info on Wiki Setup page.

**Comment by rohannexialabs** [comment]
@gotev Thanks for the message and build, will try the new release and let you know.
